src/tactic/equiv_rw.lean

```diff
@@ -0,0 +1,287 @@
+we can incrementally bootstrap the strength of `equiv_rw`.
+
+An ambitious project might be to add `equiv_rw!`,
+a tactic which, when failing to find appropriate `equiv_functor` instances,
+attempts to `derive` them on the spot.
+
+For now `equiv_rw` is entirely based on `equiv` in `Type`,
+but the framework can readily be generalised to also work with other types of equivalences,
+for example specific notations such as ring equivalence (`≃+*`),
+or general categorical isomorphisms (`≅`).
+
+This will allow us to transport across more general types of equivalences,
+but this will wait for another subsequent PR.
+-/
+
+namespace tactic
+
+/-- A hard-coded list of names of lemmas used for constructing congruence equivalences. -/
+
+-- TODO: This should not be a hard-coded list.
+-- We could accommodate the remaining "non-structural" lemmas using
+-- `equiv_functor` instances.
+-- Moreover, these instances can typically be `derive`d.
+
+-- (Scott): I have not incorporated `equiv_functor` as part of this PR,
+-- as the best way to construct instances, either by hand or using `derive`, is to use this tactic!
+
+-- TODO: We should also use `category_theory.functorial` and `category_theory.hygienic` instances.
+-- (example goal: we could rewrite along an isomorphism of rings (either as `R ≅ S` or `R ≃+* S`)
+-- and turn an `x : mv_polynomial σ R` into an `x : mv_polynomial σ S`.).
+
+meta def equiv_congr_lemmas : tactic (list expr) :=
+do exprs ←
+  [
+  `equiv.of_iff,
+  -- These are functorial w.r.t equiv,
+  -- and so will be subsumed by `equiv_functor.map_equiv` in a subsequent PR.
+  -- Many more could be added, but will wait for that framework.
+  `equiv.perm_congr, `equiv.equiv_congr, `equiv.unique_congr,
+  -- The function arrow is technically a bifunctor `Typeᵒᵖ → Type → Type`,
+  -- but the pattern matcher will never see this.
+  `equiv.arrow_congr',
+  -- Allow rewriting in subtypes:
+  `equiv.subtype_equiv_of_subtype',
+  -- Allow rewriting in the first component of a sigma-type:
+  `equiv.sigma_congr_left',
+  -- Allow rewriting ∀s:
+  -- (You might think that repeated application of `equiv.forall_congr'
+  -- would handle the higher arity cases, but unfortunately unification is not clever enough.)
+  `equiv.forall₃_congr',
+  `equiv.forall₂_congr',
+  `equiv.forall_congr',
+  -- Allow rewriting in argument of Pi types:
+   `equiv.Pi_congr_left',
+  -- Handles `sum` and `prod`, and many others:
+  `bifunctor.map_equiv,
+  -- Handles `list`, `option`, and many others:
+  `functor.map_equiv,
+  -- We have to filter results to ensure we don't cheat and use exclusively `equiv.refl` and `iff.refl`!
+  `equiv.refl,
+  `iff.refl
+  ].mmap (λ n, try_core (mk_const n)),
+  return (exprs.map option.to_list).join -- TODO: implement `.mfilter_map mk_const`?
+
+declare_trace equiv_rw_type
+
+/--
+Configuration structure for `equiv_rw`.
+
+* `max_steps` bounds the search depth for equivalences to rewrite along.
```

Comment:
> steps [...] depth

These two terms mean very different things to me.  Depth is a limit on the recursion depth, steps sounds like a limit on the number of `apply` invocations.  If this is only a depth limit, then I think it should be called `max_depth` instead of `max_steps`.

Reply:
Good point. The same applies to `solve_by_elim` itself, so if no one objects I'll make that change in this PR at the same time.